### ISSUE TITLE
Kodi doesn't save subtitle position

### ISSUE THREAD

**Issue by mkaflowski** [xbmc/xbmc#16683]
---
name: Kodi doesn't save subtitle position

---
### Describe the bug
Kodi doesn't save subtitle position in the new version. When I have change framerate change on and during movie playback go to Settings/System/Video calibration and change subtitle position it will not be saved. 

## Expected Behavior
Subtitles position in movie should be set to given value.

## Actual Behavior
It change framerate after quiting video calibration and doesnt save the values.

## Possible Fix
-

### To Reproduce
1.  Turn on "Adjust display frame rate"
2. Play some movie with different framerate/resolution that menu interface
3. Go to Settings/System/Video
4. Change subtitle position and exit

### Debuglog
-

### Screenshots 
-

## Additional context or screenshots (if appropriate)
-

### Your Environment
Used Operating system:
 - [x] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Android - Nvidia Shield
 - Kodi version: 18.4

P.S: Is there any way to save subtitle position for all resolutions and framerates?

**Comment by DaVukovic** [comment]
I can confirm this issue. But the issue only exists either with embedded subtitles or with `*.idx` subtitles. Changing the subtitle position while using `*.srt` subtitles works like a charm. 

That's probably due to the reason that `idx` files do contain some kind of fixed/forced position on the screen which is not the case for srt files. 

@mkaflowski could you please test if it works for you if you are using *.srt files

As those do work fine for me

**Comment by mkaflowski** [comment]
I use mostly srt. Changing framerate seems to clear data set.

**Comment by DaVukovic** [comment]
Then provide the log file, please

This issue might get closed without it.

In case you have problems finding it on Android, there's also an add-on available which uploads the log automatically. Search for "Kodi logfile uploader"

**Comment by MauriceW67** [comment]
I have the same issue on the Shield. Tried both the latest 18.4 nightly and the 19.x nightly

It's not just the subtitle settings that are lost, also the overscan settings.

I'll try to upload a debug log later.

Kodi forum thread: https://forum.kodi.tv/showthread.php?tid=348813

**Comment by DaVukovic** [comment]
@MauriceW67 

The screen calibration (in regard of the width and the height) shouldn't be used at all. Overscan problems should be solved at the TV itself. Try looking at the zoom options or if the TV in question does provide something like a PC mode. Correcting overscan via the screen calibration should be the very last resort. Each FullHD TV from the last 10 years should be able to natively display 1080p. Using the screen calibration is nothing more than a lost of native resolution. 

The subtitle position is another story.

**Comment by DaVukovic** [comment]
@mkortstiege 

Btw...I just tried changing the subtitle position with refreshrate-change enabled using LibreELEC and I have to admit that I can't reproduce your issue. It works like a charm while using srt-subtiltes and the position is also being saved. That works either *on-the-fly" (during playback) and also before starting the playback 

So please provide your logfile.

**Comment by MauriceW67** [comment]
Ok, I performed a completely clean install of Kodi 18.4 and reproduced the problem. Please note that I am using a nightly build (kodi-20191029-94b7cd85-Leia-arm64-v8a) because I require the fix for the Netflix addon to play 4K streams. I also tested a V19 nightly (kodi-20191028-29f64ce8-master-arm64-v8a) and it has the same issue.

Log file is from the 18.4 build: https://paste.ubuntu.com/p/B8pBnJJwG3/

Steps to reproduce (all done on an NVidia Shield TV 2017 with firmware version 8.0.1 (Android Pie)):

1. In the System menu, set Kodi resolution to 1920x1080@23.98
2. In the System menu, under Video Calibration set subtitle position to 1012 (default is 1042)
3. Play a 1920x1080@23.98 video
4. Check Video Calibration settings while playing -> All OK
5. Stop video and set Kodi resolution to to 1920x1080@59.94 in System menu
6. Play the same 1920x1080@23.98 video
7. Check Video Calibration settings while playing -> refresh rate has switched to 23.98, but subtitles are at the default of 1042 again
8. Press back button from Video Calibration screen
9. Video continues to play, but can no longer get the OSD to show by pressing OK
10. Pressing OK + cursor right and OK again stops the video, so OSD does still work, just invisible

Hope this helps!

**Comment by DaVukovic** [comment]
@MauriceW67 

So, I somewhat tested your workflow. 

Be aware that by changing the system resolution/fps, the screen calibration gets default values automatically. That also affects the subtitle position. If you change the refreshrate that should be generally the same if you change the  complete resolution. 

What I have tried here: 

- started kodi in 1920*1080
- switched to 800*600
- used the screen calibration and set some weird values
- switched back to 1920*1080
- result: screen calibration has default values

In general there's no need to switch the screen calibration/fps manually. As Kodi can do that automatically by "Adjust display refreshrate" and set that to "start/stop". That will trigger your TV to automatically switch to the matching refreshrate of the given video. You will see a black screen for about a few seconds (1-3 ... depends on how fast your TV does the switch). But by doing so, the screen calibration should stay at the values you set it to. 

I would recommend to set the Kodi GUI to 60fps. Setting it to 24fps does not look that good during navigation. 

So, from my POV, that's not a bug. That's by design. It doesn't make sense to use/copy an individual screen calibration if you change the system-resolution/fps of Kodi itself.

**Comment by DaVukovic** [comment]
@mkaflowski 

please read the above, if that's the same for you and you change the fps at the kodi settings, then I would also say, that's per design and not a bug.

**Comment by chewitt** [comment]
Kodi stores separate calibration values for each resolution. So if you have "adjust refresh" set you need to calibrate the screen values for all resolutions you will use. In other words; if you calibrate only the desktop 1080p@60 resolution and then play 1080p@23.976 media, the subs will be in the deafult position again because you didn't calibrate 1080p@23.976. Make sense?

In my personal experience when I have needed to use calibration (usually on Hotel TVs) I end up using the same calibration values on all resolutions, so it's rather annoying to repeat the calibration process 5x to cover the typical set of refresh rates for my media collection. 

It's annoying but Kodi is "working as designed" so technically it's not a bug. We should probably default to using the desktop calibrations for all refresh rates unless though.

**Comment by MauriceW67** [comment]
I think both of you don't understand what's happening on the Shield with Kodi 18.4.

What @chewitt describes above is exactly what I normally do to configure subtitle position for different resolutions and refresh rates (i.e. repeating the calibration process for each resolution and refresh rate). The issue is that these settings are no longer remembered since version 18.4.

@DaVukovic The workflow I described was just to reproduce the issue when starting from a clean install. I normally leave the GUI resolution at 60 Hz and let Kodi do the resolution and refresh rate switching for the different types of video I play.

Like I said, what I normally do is **temporarily** switch the GUI resolution to all resolutions and refresh rates to set the subtitle position for each. Then I set the GUI resolution to 60 HZ and leave it there. When you then play a video with a different refresh rate, Kodi switches to that refresh rate and uses the subtitle position that I configured previously **for that refresh rate**. This is what now no longer works. At least on the Shield. Since @DaVukovic is testing on a LibreELEC I am not sure if the bug only occurs on the Shield or on other platforms as well.

**Comment by MauriceW67** [comment]
I did some more testing and checked what is happening inside the guisettings.xml file.

I think the root cause of the bug lies in the fact that Kodi no longer reads back the values that are related to resolutions and refresh rates (such as subtitle position and overscan) from guisettings.xml.

Steps to reproduce:

1. Set Kodi GUI to a certain resolution and refresh rate (for example 1920x1080@60)
2. Go to System -> Display -> Video Calibration and change subtitle position to 1012
3. Exit the menu and then exit Kodi
4. Check value for subtitles in guisettings.xml for the 1920x1080@60 resolution, it should show 1012
5. Start Kodi again and go to System -> Display -> Video Calibration, you will see the subtitle position is back to the default of 1042 again
6. Exit the menu and close Kodi
7. Check value for subtitles in guisettings.xml for the 1920x1080@60 resolution, it now shows 1042

Looks to me like Kodi is writing the proper values to guisettings.xml, but not reading them back at startup (at least on the Shield).

**Comment by mkaflowski** [comment]
Kodi doesn't restore values for me either. It worked in the previous version.

**Comment by DaVukovic** [comment]
ok, we now have a debuglog. Still @mkaflowski refuses to provide one (for whatever reason. Obviously he's not interested in helping us to help him), so we should focus on what @MauriceW67 says as he's at least willing to be helpful. 

Thanks for that.

@MauriceW67 

I just tried: 

- before starting kodi checked the value in guisettings.xml (it was 1032)
- start kodi in 1920*1080
- change subtitle position to "981" (I guess the value doesn't matter at all as long as it's not the default one)
- closed kodi
- checked guisettings.xml -> value was still set to "981"
- started kodi again and checked the subtitle position via the GUI and it was still at 981

This is now tested on LibreELEC and Ubuntu 18.04 with the Kodi 18.4 from our PPA. So, I cannot reproduce your problem and therefor I also think it's an Android/Shield only problem. 

I'll tag this issue accordingly and we'll see if a dev jumps in.

**Comment by mkaflowski** [comment]
Sorry for no log. I am outside of home for couple days. I will upload it ASAP

**Comment by MauriceW67** [comment]
@DaVukovic 

Did some more testing to make sure. I had an old RPi2 laying around which I upgraded to LibreELEC 9.1.502 (that includes Kodi 18.4) and as you said, the issue does not happen.

I also downgraded the Kodi 18.4 on my Shield from the nightly that I was using to the latest release version from the Play Store, but the issue still occurs then.

I went through the debug log that I sent and the strange thing is that it does not log anything about loading guisettings.xml, nor does it report anything about setting subtitle position or overscan settings. Not sure if that is to be expected?

**Comment by DaVukovic** [comment]
I'm not sure about that either. 

Another Team member who also owns a Shield unfortunately also can't reproduce your issue. For him it seems that the specific value is stored and also read from at the next Kodi start.

**Comment by mkaflowski** [comment]
There is a movie with bug: [https://photos.app.goo.gl/7aCvpH8Ai6NEJtwF9](url). You can see it at 0:40. When I go back the framerate is changed and values are gone.
Logs:
[https://pastebin.com/wh8iE8Ck](https://pastebin.com/wh8iE8Ck)

**Comment by mkaflowski** [comment]
Second video: [https://photos.app.goo.gl/gsByeeEZJ5HHZhET8](https://photos.app.goo.gl/gsByeeEZJ5HHZhET8)

**Comment by MauriceW67** [comment]
Seems that no developer is willing to take a look at this issue sofar.

In the mean time I did some further testing on my Shield:

- I completely uninstalled Kodi and installed a fresh copy of the release version of Kodi 18.2 (arm64-v8a)
- Tested overscan and subtitle settings while switching between different refresh rates -> All settings retained
- I installed Kodi 18.3 release version over 18.2
- Tested overscan and subtitle settings while switching between different refresh rates -> All settings retained
- I installed 18.4 release version over 18.3
- Tested overscan and subtitle settings while switching between different refresh rates -> Settings no longer retained!

So it looks like some kind of regression occured somewhere between version 18.3 and 18.4 that breaks reading back the overscan and subtitle position settings from guisettings.xml.

Is there an easy way to see which changes were applied between 18.3 and 18.4 through Github? Maybe if there are not too many it would be possible to pinpoint where things went wrong?

**Comment by olivluca** [comment]
If you look here https://forum.kodi.tv/showthread.php?tid=310512 you'll see that this has been happening on and off since 2017. I'm currently having this problem (calibration not saved) with kodi 18.4 on raspbian and on the beta of libreelec with kodi 19. I'm only using one resolution and one refresh rate because I'm trying to use the analog video output with an old tv set.

**Comment by olivluca** [comment]
I took a look at [DisplaySettings.cpp](https://github.com/xbmc/xbmc/blob/aa39210197323230c1c2ab086a6f2a18f39daf9a/xbmc/settings/DisplaySettings.cpp)

It has a method Load that supposedly loads the overscan settings, however

1. I couldn't find a method to save those settings (I just looked at that class, not its ancestors)
2. I don't see that method called anywhere (I just used github search so I could have missed it)

I tried to manually put the following <resolutions> section in guisettings.xml, replacing the empty one that was there, with no effect

```xml    <resolutions> 
       <resolution> 
           <description>720x576@50.00</description>
           <subtitles>517</subtitles>
           <pixelratio>1.034</pixelratio>
           <overscan> 
             <left>18</left>
             <top>53</top>
             <right>14</right>
             <bottom>34</bottom>
           </overscan>
       </resolution>
    </resolutions>
```

**Comment by MauriceW67** [comment]
@olivluca I tried manually setting values in guisettings.xml before (see Kodi forum thread), but they are just being ignored.

Somehow the behavior changed between Kodi 18.3 and 18.4 on Shield/Android, but I don't have enough Github and programming knowledge to find out what happened.

Forum thread: https://forum.kodi.tv/showthread.php?tid=348813

**Comment by olivluca** [comment]
[these are the changes](https://github.com/xbmc/xbmc/compare/18.3-Leia...18.4-Leia#diff-fdf8a0ecba6648be7362d3a49502479c) between 18.3 and 18.4.
The only change that I see that could affect this is the change in DisplaySettings.cpp (click on the files tab to see it), but since I know nothing about kodi architecture I could very well be wrong.

**Comment by DaVukovic** [comment]
For everyone in here. Please stop those "me too" spammings without providing logs. Links to forum threads where no logs are included also doesn't help anyone. 

@mkaflowski the last log you provided is "again" not a debuglog. 

It also seems you have a refreshrate change each time you exit the video. Please try to set the refreshrate change only on start/stop of the video and see if that helps .

**Comment by mkaflowski** [comment]
@DaVukovic It is set to "on start/on stop".

**Comment by DaVukovic** [comment]
ok, then I'm seriously wondering why your screen goes black after you finished the video calibration at the first glance. 

Will have to check that at home as my current PC montior doesn't allow any kind of refreshrate-change at all

**Comment by olivluca** [comment]
I just wanted to point out that it's not only the subtitle position that isn't stored/reloaded but all the settings that supposedly should be stored under ```<resolutions>``` and it's not android specific.
Is it possible/desirable to change the title?

**Comment by MauriceW67** [comment]
@DaVukovic I think @mkaflowski is changing the video calibration settings in the wrong place in his video. He changes it while playing the video and going to the Video settings -> Video calibration settings through the OSD. AFAIK, this will only change the calibration settings for the current video. This will indeed make the screen go blank for a second when exiting the calibration settings.

The correct way is as I described: Going to the Video calibration settings under System -> Display from the Kodi settings menu. Then (temporarily) changing to a particular resolution and refresh rate and change the video calibration settings. Then the next time you play a video with the same resolution and refresh rate, the new settings should be used. But since version 18.4 they are not, which is the core issue in this bug report.

**Comment by MauriceW67** [comment]
Any news on this? If a new log is required or I can help by running a debug build, let me know.

**Comment by fritsch** [comment]
@ksooo https://github.com/xbmc/xbmc/pull/16340 <- could you take a look? Seems your PR does something in that regard. Thanks much.

**Comment by MauriceW67** [comment]
@ksooo Thanks to @fritsch who pointed me to #16340 

I went through the trouble of manually reverting the changes in #16340 and recompiled Kodi on my local build system.

And indeed applying and retaining the calibration settings and subtitle positions is now working again between resolution and refresh rate changes.

I hope this helps in fixing this bug.

**Comment by MauriceW67** [comment]
@ksooo Could you take a look please?

On the forum I already did some debugging, perhaps that will help in narrowing down where things go wrong:

https://forum.kodi.tv/showthread.php?tid=348813&pid=2910113#pid2910113

**Comment by mkaflowski** [comment]
Any progress on that?

**Comment by MauriceW67** [comment]
Unfortunately no progress sofar.

@fritsch has identified which PR has caused the regression, but the developer who coded it (@ksooo) has not looked into it yet AFAIK.

**Comment by ksooo** [comment]
@MauriceW67 @mkaflowski  could you please try this test build and report back? It is latest Leia nightly plus the potential fix for the regression you guys have reported. => https://jenkins.kodi.tv/job/Android-ARM64/13189/artifact/kodi-20200131-36efa54c-leia-android-fix-updateresolutions-arm64-v8a.apk

**Comment by mkaflowski** [comment]
Thank you. Will do tomorrow.

pt., 31 sty 2020, 22:14 użytkownik Kai Sommerfeld <notifications@github.com>
napisał:

**Comment by MauriceW67** [comment]
@ksooo thanks a lot for this build. I just tested it and both subtitle position and overscan settings are being saved and restored properly again!

**Comment by ksooo** [comment]
@MauriceW67 thanks for the feedback. Glad to hear it's working again.

**Comment by mkaflowski** [comment]
Works! thank you :)

**Comment by brozikcz** [comment]
It will be available the update with this fix via the Google Play?

**Comment by olivluca** [comment]
I see that the patch is only for android. No joy for the raspberry pi?

**Comment by MauriceW67** [comment]
@brozikcz The fix is already in the nightlies of Kodi Leia 18.6:

http://mirrors.kodi.tv/nightlies/android/

So it will also be in Kodi Leia 18.6 final when released.

@olivluca During my tests I also tested with LibreELEC 9.x on a Raspberry Pi 2 and I could not reproduce the bug, so if you have a problem it must be caused by something else.

**Comment by djhifi** [comment]
I can also confirm this bug (with adjust display refresh rate ON) and furthermore i've tried setting up subtitle position for every framerate available and it always defaults back to default value whenever I play the media. If I try to change position when video is playing (to then "set default for all media"), it freezes the remote control functions and the "ok" Nvidia shield button is then used as a pause/play (so I cannot access OSD again). Will try the nightly.
